packages/peregrine/src/List/list.js

```diff
@@ -9,89 +9,84 @@ import Items from './items';
-class List extends Component {
-    static propTypes = {
-        /**
-         * Class names to use when styling this component
-         */
-        classes: PropTypes.shape({
-            root: PropTypes.string
-        }),
-        getItemKey: PropTypes.func.isRequired,
-        /**
-         * An iterable that yields `[key, item]` pairs such as an ES2015 Map
-         */
-        items: iterable.isRequired,
-        /**
-         * A render prop for the list element. A tagname string, such as `"div"`, is also valid.
-         */
-        render: PropTypes.oneOfType([PropTypes.func, PropTypes.string])
-            .isRequired,
-        /**
-         * A render prop for the list item elements. A tagname string, such as `"div"`, is also valid.
-         */
-        renderItem: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
-        /**
-         * A callback that fires when the selection state changes.
-         */
-        onSelectionChange: PropTypes.func,
-        /**
-         * A string corresponding to a selection model.
-         */
-        selectionModel: PropTypes.oneOf(['checkbox', 'radio'])
-    };
-
-    static defaultProps = {
-        classes: {},
-        getItemKey: ({ id }) => id,
-        items: [],
-        render: 'div',
-        renderItem: 'div',
-        selectionModel: 'radio'
-    };
-
-    render() {
-        const {
-            classes,
-            getItemKey,
-            items,
-            render,
-            renderItem,
-            onSelectionChange,
-            selectionModel,
-            ...restProps
-        } = this.props;
-
-        const customProps = {
+function List(props) {
+    const {
+        classes,
+        getItemKey,
+        items,
+        render,
+        renderItem,
+        onSelectionChange,
+        selectionModel,
+        ...restProps
+    } = props;
+    const customProps = useMemo(
+        {
             classes,
             getItemKey,
             items,
             onSelectionChange,
             selectionModel
-        };
-
-        const Root = fromRenderProp(render, Object.keys(customProps));
-
-        return (
-            <Root className={classes.root} {...customProps} {...restProps}>
-                <Items
-                    items={items}
-                    getItemKey={getItemKey}
-                    renderItem={renderItem}
-                    selectionModel={selectionModel}
-                    onSelectionChange={this.handleSelectionChange}
-                />
-            </Root>
-        );
-    }
-
-    handleSelectionChange = selection => {
-        const { onSelectionChange } = this.props;
-
-        if (onSelectionChange) {
-            onSelectionChange(selection);
-        }
+        },
+        [classes, getItemKey, items, onSelectionChange, selectionModel]
+    );
+    const Root = useMemo(fromRenderProp(render, Object.keys(customProps)), [
+        render,
+        customProps
+    ]);
+    const handleSelectionChange = selection => {
```

Comment:
Will `onSelectionChange` not get a new reference on every render?

Reply:
Maybe, but the benefit is that `handleSelectionChange` will be memoized based on the reference so you can optimize up the tree to wherever/whatever is making `onSelectionChange` different.

Reply:
Sure. Will memoize it using `useCallback` but we should be doing this with an assumption that the parent whoever is using this component will memoize `onSelectionChange` before passing it down to the `List` component as prop. If not, we will be assigning more work to the garbage collector since now it has to clean up 2 functions instead of 1 on every render.